compiler/rustc_query_system/src/dep_graph/serialized.rs

```diff
@@ -135,8 +133,30 @@ impl<'a, K: DepKind + Decodable<MemDecoder<'a>>> Decodable<MemDecoder<'a>>
             debug_assert_eq!(_i.index(), _index);
         }
 
-        let index: FxHashMap<_, _> =
-            nodes.iter_enumerated().map(|(idx, &dep_node)| (dep_node, idx)).collect();
+        let mut duplicates = Vec::new();
+        let mut index: FxHashMap<DepNode<K>, SerializedDepNodeIndex> = Default::default();
+        index.reserve(nodes.len());
+        for (idx, &dep_node) in nodes.iter_enumerated() {
+            if index.insert(dep_node, idx).is_some() {
+                duplicates.push(dep_node);
+            }
+        }
+
+        // Creating the index detected a duplicated DepNode.
+        //
+        // If the new session presents us with a DepNode among those, we have no
+        // way to know which SerializedDepNodeIndex it corresponds to. To avoid
+        // making the wrong connection between a DepNodeIndex and a SerializedDepNodeIndex,
+        // we remove all the duplicates from the index.
+        //
+        // This way, when the new session presents us with a DepNode among the duplicates,
+        // we just create a new node with no counterpart in the previous graph.
+        //
+        // Red/green marking still works for those nodes, as that algorithm does not
+        // need to know about DepNode at all.
+        for dep_node in duplicates {
+            index.remove(&dep_node);
+        }
```

Comment:
There might be some logic which assumes `SerializedDepNodeIndex` refers to an unique `DepNode`. It may be less risky to just reject the incremental cache when a collision is detected.

Reply:
You can still observe that different dep node indices refer to the same dep node using `SerializedDepGraph::index_to_node`, so it's not contained to this module.